Integrate UI tests sharding

wzieba opened the pull request:
<!-- Remember about a good descriptive title. -->
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR introduces instrumented tests sharding. Instead of running all tests on a single device, Fladle Gradle Plugin (Flank under the hood) divides tests into groups that expected execution of all is no longer than 2 minutes and runs all tests simultaneously, reducing the overall waiting time.

| Before | After | 
| --- | --- |
| <img width="937" alt="image" src="https://github.com/user-attachments/assets/d67b0b13-d955-4560-9c98-254cf37be2a2"> | ![image](https://github.com/user-attachments/assets/8e2da5d6-75f9-4dbb-91fd-b0d08118f56f) |

#### How does Flank now how to group tests?

It uses historical data of tests execution. This data is saved at gs://fladle/woocommerce-android/smart-flank/JUnitReport.xml

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

Green CI checks should be fine.


- [ ] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

wzieba committed:
Integrate and configure Fladle Gradle Plugin

wzieba committed:
Set localResultDir to the one that is defined in `run-instrumented-tests.sh`

So we can report tests results to Buildkite

wzieba committed:
Run Flank instead `android_firebase_test`

Most of the removed configuration is moved to `flank{}` Gradle extension.

The removed logic for annotating Buildkite job is intentional. For now, I'd like to keep the configuration minimal. Later we can introduce annotating, if necessary, by taking details_url from flank build reports

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/226">19.6</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

wzieba committed:
Merge branch 'disable_tablet_tests' into integrate_ui_tests_sharding

wzieba committed:
Remove specifying UI test device parameters

It's now declared in Gradle's `fladle{}` extension

wzieba committed:
Print all apk files to debug missing test apk reported by Flank

wzieba committed:
Copy the whole matrix directory. Run flank in the same gradle invocation

wzieba committed:
Run flank as a separate Gradle invocation

Otherwise, error raises on CI: `automattic/woocommerce-android/WooCommerce/build/outputs/apk/androidTest/vanilla/debug/WooCommerce-vanilla-debug-androidTest.apk' from test doesn't exist`. It seems to be some issue with Fladle plugin

wzieba commented on fastlane/Fastfile:
```diff
@@ -967,35 +967,7 @@ platform :android do
-      crash_on_test_failure: false
-    )
-
-    annotation_ctx = 'firebase-test-woocommerce-vanilla-debug'
-    if test_succeeded
-      sh("buildkite-agent annotation remove --context '#{annotation_ctx}' || true") if is_ci?
-    else
-      details_url = lane_context[SharedValues::FIREBASE_TEST_MORE_DETAILS_URL]
-      message = "Firebase Tests failed. Failure details can be seen [here in Firebase Console](#{details_url})"
-      sh('buildkite-agent', 'annotate', message, '--style', 'error', '--context', annotation_ctx) if is_ci?
-      UI.test_failure!(message)
-    end
```
I decided to drop the annotation logic for now. It's a nice addition, but I'm not sure if crucial, and I haven't seen a straightforward way to obtain the matrix URL. If you think otherwise and it's best to keep this annotation, please ping me and I'll work on it.

ParaskP7 replied:
>  It's a nice addition, but I'm not sure if crucial...

Not crucial definitely, but very helpful nevertheless.

> ...I haven't seen a straightforward way to obtain the matrix URL.

Yea, I am not sure as well, I actually asked you about this, maybe related... 🤷 

> I decided to drop the annotation logic for now.
> If you think otherwise and it's best to keep this annotation, please ping me and I'll work on it.

👍 I also think this shouldn't block this PR, definetely not, but I would suggest not forgetting to add this annotation logic back, maybe creating an accompanying issue on it, not sure. 🤔

wzieba committed:
Merge branch 'trunk' into integrate_ui_tests_sharding

wzieba committed:
Address `Unused block argument - `options`.` Ruby lint warning

jostnes commented on WooCommerce/build.gradle:
```diff
@@ -13,6 +13,23 @@ plugins {
+    testTargets = [
+            "notPackage com.woocommerce.android.e2e.tests.screenshot"
+    ]
+    devices = [
+            [ "model": "Pixel2.arm", "version": "30" ]
+    ]
```
If/when the tablet-instrumented tests are reintroduced, how can we get those tests to be part of this, too? Do we include the tablet emulator as part of the devices list? How would the report work?

wzieba replied:
Yes, you're right - we'd have to add tablet to devices here. 

I've prepared an example: 0d5ebdc and then report looks like this https://buildkite.com/organizations/automattic/analytics/suites/woocommerce-android-instrumented-tests/runs/499097ee-0404-81fb-b1d7-6efd49ede403

wzieba replied:
or, more detailed, at Firebase console https://buildkite.com/automattic/woocommerce-android/builds/22460#0190c4bf-7469-45e7-ae1b-064bb418f40d/269-707

wzieba replied:
I'll go ahead and merge this PR, but I'm happy to continue this thread! (here or on Slack)

jostnes replied:
Thank you for confirming and sharing the links! Glad that it's this easy to add new devices!

Although we can differentiate devices in FTL, Buildkite's Test Analytics combines the results from both phone and tablet into one test suite (it's one step on the build), making it difficult to identify the device of a failing test when it happens. 

I'm not sure how big of an issue this will be, though (since we could still view the results from FTL). But do you know if it is possible to add the device type to be displayed on the test case name (I'm not sure how the test name is generated)?

wzieba replied:
Thanks @jostnes , good points!

> But do you know if it is possible to add the device type to be displayed on the test case name (I'm not sure how the test name is generated)?

I'll think how we can address this, maybe there's some way to add more details for Buidlkite Test collector! Maybe we can go back to this topic when we'll be enabling tablet tests again?

wzieba committed:
temp: add tablet as a test device

wzieba committed:
Revert "temp: add tablet as a test device"

This reverts commit 0d5ebdcba5b1cfb0d493d5ab3664a5149cdfa18e.

ParaskP7 commented on .buildkite/commands/run-instrumented-tests.sh:
```diff
@@ -32,6 +28,6 @@ else
 
 echo "--- 🧪 Copying test logs for test collector"
-mkdir WooCommerce/build/buildkite-test-analytics && cp build/instrumented-tests/*/*/*.xml WooCommerce/build/buildkite-test-analytics
+mkdir WooCommerce/build/buildkite-test-analytics && cp -r build/instrumented-tests/matrix_* WooCommerce/build/buildkite-test-analytics
```
Question (❓): Can you help me understand what coping the whole matrix directory will look like and why we need that? 🙏

iangmaia replied:
+1 -- I don't quite what generated this matrix and how it's used by Fladle (thinking about how it could be accessible by the builds as discussed on https://github.com/woocommerce/woocommerce-android/pull/12029#discussion_r1680631252).

wzieba replied:
The structure of files is different because we have multiple shards, instead of a single execution

<img width="435" alt="Screenshot 2024-07-18 at 12 41 23" src="https://github.com/user-attachments/assets/f011dd53-6d1d-4705-9443-a2130d6937ba">

iangmaia replied:
Ah, got it, I thought this was the input to the builds and it's the test output instead, right?
So it seems Fladle input is just the `devices` array in `build.gradle`?

wzieba replied:
Yes, this is test output, that is later collected by Buildkite.

> So it seems Fladle input is just the devices array in build.gradle?

I mean - everything in `fladle { }` extension in `build.gradle` is input for Fladle, including the `devices` array.

ParaskP7 commented on fastlane/Fastfile:
```diff
@@ -964,38 +964,10 @@ platform :android do
   #####################################################################################
   desc 'Build the application and instrumented tests, then run the tests in Firebase Test Lab'
-  lane :build_and_instrumented_test do |options|
+  lane :build_and_instrumented_test do |_options|
```
Minor (🔍): Since we are no longer using `options` for this lane, maybe we should instead remove this altogether, just to make this lane as straightforward as possible, wdyt?

wzieba replied:
It was `_options` before recent changes
https://github.com/woocommerce/woocommerce-android/pull/11829/files#diff-dff4b99d4e651ab9d7597876ec8dbe92aa934e42c0fb55f4aadd6c106fc96688L967

@AliSoftware or @iangmaia is `_options` okay or should I remove this parameter (if it's possible at all?)?

ParaskP7 replied:
Yea it was `options` before because of the below usage:

```
      model: options[:device],
      version: options[:version],
      orientation: options[:orientation],
```

iangmaia replied:
Exactly @ParaskP7, it was used before and I guess RuboCop renamed it to `_options` because it is not used anymore?

Anyway, it should be safe to remove it:

```suggestion
  lane :build_and_instrumented_test do
```

wzieba replied:
👍 addressed in c9ad056ae32cf3536b7474b79f8a9ec080adc4b7

ParaskP7 commented on fastlane/Fastfile:
```diff
@@ -964,38 +964,10 @@ platform :android do
-      sh('buildkite-agent', 'annotate', message, '--style', 'error', '--context', annotation_ctx) if is_ci?
-      UI.test_failure!(message)
-    end
+    gradle(task: 'runFlank')
```
Praise (❤️): So much configuration is gone now that we are using `fladle` and pointing to `serviceAccountCredentials` instead, love it!

ParaskP7 reviewed (approved):
👋 @wzieba !

I have reviewed this PR and everything LGTM, a really awesome job done here, kudos! 🌟 x 🌟 ^ 🌟 

1. It is awesome that with this change now we are seeing that what used to take about `38m`, it is now taking `14m` instead, this is HUGE! 💯 
2. I also checked the before and after on the Firebase side, without the shards, it took about `10m` to run all `57` test cases, and now it takes the `2m` max per shard (as configured), 6 shards in total, to run all those test cases.
3. I also noticed that the 1st shard run 49/57 test cases. Then the other 5 shards run 1-to-2 test cases max. I wonder if that is indeed expected and `fladle->flank` is actually that smart to read the historical report, understand that those 49 are very quick, put as many into that shard and get it done. And vice-versa, for it to be able to notice that all the other 8/57 test cases is so slow that they need (almost) their own shard to run. Interesting... 💭 

PS: Although I have been hearing and wanting to try `fradle->flank` myself for quite some time now, I was always wondering about its maintainability going forward. [Nelson](https://github.com/runningcode) is not some random guy, and he has been working with Gradle for quite some time now, but would he be able to support [fladle](https://github.com/runningcode/fladle/) indefinitely for us and all our projects? For example, before the last [0.17.5](https://github.com/runningcode/fladle/releases/tag/v0.17.5) release (`13/06/24`), the previous one was almost 2 years ago (`29/05/22`). On the other hand, to my understanding [fladle](https://github.com/runningcode/fladle/) is kind of a facet on top of [flank](https://github.com/Flank/flank), which is run by an actual organization, so this seems more maintainable to me. Their [simple-frank](https://github.com/Flank/simple-flank) Gradle plugin got last updated a year ago (`12/05/2023`), I guess this is fine, right? Maybe that tool doesn't need to many updates to integrate nicely with Firebase Test Lab, not sure. Anyway, I am sure we are doing a great move here, I just wanted to mention the only limitation I see with switching from vanilla Firebase to `fladle->flank`.

-----

I have left a question (❓) and a minor (🔍) comment for you to consider. I am going to approve this PR anyway, since none is blocking. I am **NOT** going to merge this PR yet to give you some time to apply any of my suggestions. However, feel free to ignore them and merge the PR yourself.

ParaskP7 commented:
FYI: By the way, I didn't go through all the [configuration](https://runningcode.github.io/fladle/configuration/) available on `fladle`, and there are lots, so I trust that what we have now is enough, meaning, enough to get us going and migrate from vanilla Firebase to `fladle`. Maybe the next step is to try and optimize `fladle`, per project, and see if we can optimize the time-to-completion even further.

wzieba committed:
Remove unused `_options`

wzieba commented:
Thanks for the review @ParaskP7 !

> I wonder if that is indeed expected and fladle->flank is actually that smart to read the historical report

Yes, it's called `smartFlank` and the result of each build is published at https://console.developers.google.com/storage/browser/fladle/woocommerce-android/smart-flank/JUnitReport.xml

>  I was always wondering about its maintainability going forward

In case if this plugin will be abandoned, we can always configure `flank` ourselves. Flank doesn't look any close to be abandoned, but if it would happen, we can simply revert to use Firebase Test Lab directly.

ParaskP7 commented:
> Yes, it's called smartFlank and the result of each build is published at https://console.developers.google.com/storage/browser/fladle/woocommerce-android/smart-flank/JUnitReport.xml

Yea, I know, I was just wondering how smart-smart that it, seeing this `49/57` on 1 shard vs. `8/57` on 5 shards and all. To be honest, I am quite impressed! 💯 

> In case if this plugin will be abandoned, we can always configure flank ourselves. Flank doesn't look any close to be abandoned, but if it would happen, we can simply revert to use Firebase Test Lab directly.

True, let's just hope this will continue to work for at least 5 more years, until Google acquire `flank` and add that process in-house, making us pay for it too (of course)... 🤣 

On the other hand, us reverting to FTL and doubling out UI test execution times seems frightening to me, when we make improvements everyone is happy, but the other way around is quick bitter... 😅 🤞
